nbgrader/exchange/ngshare/exchange.py

```diff
@@ -18,10 +19,60 @@
 
 class Exchange(ABCExchange):
 
-    ngshare_url = 'http://172.17.0.3:11111'  # need to get IP address of container
-    username = os.environ['USER'] 
+    username = os.environ['JUPYTERHUB_USER']
+
+    @property
+    def ngshare_url(self):
+        if 'PROXY_PUBLIC_SERVICE_HOST' in os.environ:
+            # we are in a kubernetes environment, so dns based service discovery should work
+            # assuming the service is called ngshare, which it should
+            return "http://proxy-public/services/ngshare" + self.prefix
+        else:
+            # TODO: maybe expose this in the nbgrader configs?
+            # for now, keeping the original url to not break docker testing setup
+            return 'http://172.17.0.3:11111' + self.prefix  # need to get IP address of container
```

Comment:
`self.prefix` being referenced before `prefix`?

Reply:
It shouldn't be an issue, since it's a function that will get called whenever someone references `ngshare_url`, and by that point `prefix` is evaluated. That said, apparently the prefix is different between vngshare and ngshare, so I just removed the prefix and coded them in `ngshare_url` instead.